Initial tagging system for MooseDocs

cticenhour opened the pull request:
Refs #25387

Still needs some documentation and probably a bit more clean-up, but putting it up for testing and review for those interested.

I would consider this draft of the extension very experimental. It currently replaces data in a user-provided JavaScript file with the combined tag content that it finds when the extension command is triggered. A mechanism to have the filtering page available to the rendered website still needs to be created, so this PR doesn't close the original issue. 

Tagging @GiudGiud for awareness. We can discuss next steps when we're back to work on Tuesday. 

@aeslaughter, as we discussed and if you are interested, feel free to look at and review how @DempseyRogers1994 and I put together the extension.

DempseyRogers1994 committed:
Added the first version of the tagger and tags.pklto the extensions.

Refs #25387

DempseyRogers1994 committed:
Changed output to .txt and added name and path to dict

edited notes in header

DempseyRogers1994 committed:
Reduced the amount of Input

DempseyRogers1994 committed:
added appendable dict()

DempseyRogers1994 committed:
adding test_tagging.py renamed tagger-->tagging

cticenhour committed:
Add log to tagging extension

cticenhour committed:
Remove whitespaces from tagging extension source

DempseyRogers1994 committed:
Serial and Push to Static

DempseyRogers1994 committed:
More "reactor-y" tags

cticenhour committed:
Remove more trailing whitespaces from tagging extension

cticenhour committed:
Send tagging extension to end of default list

cticenhour committed:
Add tagging test spec

cticenhour committed:
Clean up tagging test template

cticenhour committed:
Clean up tagging extension frontmatter

cticenhour committed:
Allow allowed_keys to be set in the config.yml file

cticenhour committed:
Improve output messages for tagging extension

cticenhour committed:
Fix typos in MooseDocs comments

cticenhour committed:
Remove tagging extension from default list

Apps should only opt-in when necessary

cticenhour committed:
Re-organize tagging extension using global attributes

This preserves parallelism in createToken and pushes find-and-replace into the postExecute method.

Refs #25387

cticenhour committed:
Add experimental notices and remove un-used extension iterator method.

cticenhour committed:
Add test pages for tagging extension

cticenhour committed:
Improve tagging extension warning messages

cticenhour committed:
Fix typo

cticenhour committed:
Ensure that iteration through global attributes for tagging extension only takes in attributes created by the tagger using a unique string prefix

cticenhour committed:
Add context and linking to markdown test pages.

cticenhour committed:
Switch to generated new file instead of modifying original file in place

This current version causes an endless loop of post-execution as the generated file is updated. An approach similar to the media extension script capability (pre-allocation and copying direct to destination) should likely be explored.

cticenhour committed:
Begin setting up tagging extension testing

cticenhour committed:
Fix error in tag page identifier check

Also shift postExecute to be after get/set methods and allowed_keys property

cticenhour committed:
Add extra context to third tagging test page

cticenhour committed:
Begin writing directly to destination during postExecute and fixup incorrect replace_str

No more endless write loop!

cticenhour committed:
Add tagging extension initialization test

cticenhour committed:
Add check for duplicate keys

cticenhour committed:
Simplify initialization test and add command testing for tagging extension

cticenhour committed:
Add check and error for no key:value pairs

cticenhour committed:
Add explicit testing for duplicate names

Remove unintentional case from general command testing

cticenhour committed:
Add Tag token to tagging extension

cticenhour committed:
Add 'name' guess when there is a syntax spacing issue between the tagger command a possible name

cticenhour committed:
Consolidate duplicate names test into command testing series

cticenhour committed:
Simplify log messages in tagging extension

Fixup typo in load_config default extensions

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1747491/) on d1b6070 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25388/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1747534/) on d1b6070 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>395233</th>
      <th colspan=4>#25388 d1b607</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.66%</td>
      <td>85.66%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90538</td>
      <td>90533</td>
      <td>-5</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15154</td>
      <td>15159</td>
      <td>+5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25388/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25388/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25388/coverage/xfem)

</details>



This comment will be updated on new commits.

cticenhour committed:
Transition from 'tagger' --> 'tag' in command

Also update heading in-source documentation

cticenhour committed:
Set tagging extension subcommand to None and use name settings option instead

This fixes up verify testing errors for HTML and LaTeX. Also enable tagging extension in both of those yaml config files.

cticenhour committed:
Move to extension settings parameters for name and pairs

This improves error checking and simplifies syntax considerations. Remove second and third testing markdown pages - not necessary anymore.

cticenhour committed:
Add tagging extension documentation

cticenhour committed:
Add tagview static content

cticenhour committed:
Turn on tagging extension in the electromagnetics module

cticenhour committed:
Add tags to electromagnetics benchmarks and verification cases

cticenhour committed:
Fixup MooseDocs test specification issues

- Modifies a potential typo (13567 becomes 13580)
- Adds previous issues that were missed with new unit tests
- Moves tagging extension issue to top-level 'issues' parameter

GiudGiud commented on python/MooseDocs/extensions/tagging.py:
```diff
@@ -0,0 +1,189 @@
+ALERT: The tagging extension is experimental! If documentation tagging features are desired, please
+request assistance from the MOOSE Framework development team.
+
+Example Tagger command in *.md:
```
```suggestion
Example Tag command in *.md:
```

GiudGiud commented on python/MooseDocs/extensions/tagging.py:
```diff
@@ -0,0 +1,189 @@
+        """
+        replace_str = ""
+        for iter in self.getAttributeItems():
+            if bool(re.search('tagger_', iter[0])):
```
```suggestion
            if bool(re.search('tag_', iter[0])):
```

GiudGiud commented on python/MooseDocs/extensions/tagging.py:
```diff
@@ -0,0 +1,189 @@
+            page_data = {'name':name, "path":mpath, "key_vals":dict(good_keys)}
+
+            tag_id_name = ''
+            if self.extension.get_tag_data("tagger_" + name):
```
```suggestion
            if self.extension.get_tag_data("tag_" + name):
```

GiudGiud commented on python/MooseDocs/extensions/tagging.py:
```diff
@@ -0,0 +1,189 @@
+                    "dictionary: %s"
+                LOG.warning(msg, page.name, name)
+            else:
+                tag_id_name = "tagger_" + name
```
```suggestion
                tag_id_name = "tag_" + name
```

GiudGiud commented on python/MooseDocs/test/extensions/test_tagging.py:
```diff
@@ -0,0 +1,100 @@
+        with self.assertLogs(level=logging.INFO) as cm: # For warning suppression. TODO: remove cm when non-experimental
+            ast = self.tokenize(self.TEXT)
+        self.assertSize(ast, 1)
+        self.assertEqual(ast(0)['attr_name'], 'tagger_test')
```
```suggestion
        self.assertEqual(ast(0)['attr_name'], 'tag_test')
```

GiudGiud commented on python/MooseDocs/test/extensions/test_tagging.py:
```diff
@@ -0,0 +1,100 @@
+        self.assertEqual(len(cm.output), 2)
+        self.assertIn('Tag page identifier already exists;', cm.output[1])
+        self.assertSize(ast1, 1)
+        self.assertEqual(ast1(0)['attr_name'], 'tagger_test')
```
```suggestion
        self.assertEqual(ast1(0)['attr_name'], 'tag_test')
```

GiudGiud commented on python/MooseDocs/extensions/tagging.py:
```diff
@@ -0,0 +1,189 @@
+    @staticmethod
+    def defaultSettings():
+        settings = command.CommandComponent.defaultSettings()
+        settings['name'] = (None, 'ID name for page and associated key:value category/label pairs.')
```
```suggestion
        settings['name'] = (None, 'ID name for page and associated key:value category:label pairs.')
```

GiudGiud commented on python/MooseDocs/extensions/tagging.py:
```diff
@@ -0,0 +1,189 @@
+                msg = "%s: Provided 'key' not in allowed_keys (see config.yml); not adding the " \
+                       "following to the database: %s"
+                LOG.warning(msg, page.name, pair[0])
+            elif len(good_keys) != 0 and pair[0] in good_keys[0]:
```
this only checks the first entry.

I think you want to check all pairs for their first entry
not sure what the one-liner is for that

cticenhour replied:
Oh good catch. This was working "well enough" that I didn't catch it in my local testing. I think the fix is to simply perform `pair[0] in good_keys` for that part of the check and Python is smart enough to look through all without extra fuss. Will try that first.

cticenhour replied:
Fixed in e916415

GiudGiud commented on python/doc/content/python/MooseDocs/extensions/tag.md:
```diff
@@ -0,0 +1,54 @@
+# Tagging Extension
+
+The tagging extension provides the "tag" command, which provides a mechanism for setting a set of
+tags made up of arbitrary key:value pairs for categories/labels associated with a given markdown page.
```
```suggestion
tags made up of arbitrary key:value pairs for categories:labels associated with a given markdown page.
```

GiudGiud commented on python/doc/content/python/MooseDocs/extensions/tag.md:
```diff
@@ -0,0 +1,54 @@
+
+There is only one version of the tag command, demonstrated below. The available settings for the
+tag command are listed in [tagging-settings].
+
```
say what is tagging_one (the name) here

cticenhour replied:
Done in b2ee9a4

GiudGiud commented on python/doc/content/python/MooseDocs/extensions/tag.md:
```diff
@@ -0,0 +1,54 @@
+documentation tagging features are desired, please request assistance from the MOOSE Framework
+development team.
+!alert-end!
+
```
is there a page online for MOOSE that we can link to? that shows this active for a few things?

cticenhour replied:
There is not with this PR - I could add one to the electromagnetics module (move a couple commits from my other branch to this one) if you want a sample that will be online regularly. Adding it to the test dirs for MooseDocs wouldn't end up being published online.

cticenhour replied:
Once the doco rebuilds, an example page should be live at https://mooseframework.inl.gov/docs/PRs/25388/site/modules/electromagnetics/filter/index.html

GiudGiud replied:
seems it s not working

cticenhour replied:
Looks like an issue with the HTML. Browser is looking in `https://mooseframework.inl.gov/js/index-93b559a6.js`, which obviously doesn't exist. [The actual location](https://mooseframework.inl.gov/docs/PRs/25388/site/js/index-93b559a6.js) does however and contains the right information. Will need to explore this tomorrow.

cticenhour replied:
Good to go now at the link I posted earlier. Moved to manual relative paths within the HTML, consistent with others that MooseDocs generates.

GiudGiud replied:
so the filters arent quite combining the way we wanted.

for example, if you specify benchmark and 2021, we would expect to see less results (only the benchmarks from 2021), not the benchmarks + the models contributed in 2021 (so more results)

cticenhour replied:
That would be a @friedmud question. The filtering functionality itself is within his tagview repository and outside of this PR.

cticenhour replied:
Also I agree with what you are saying - I would also like to see further filtering with more categories selected.

cticenhour committed:
Apply initial suggestions from code review

Refs #25388

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

aeslaughter commented:
@cticenhour I won't have time to look at this for a little while, so don't wait for me.

cticenhour committed:
Address review comment: fixup duplicate key check bug in tagging extension and add complexity to unit test

Refs #25388

cticenhour committed:
Address review comment: add more content to tagging extension doco

Refs #25388

cticenhour committed:
Shift to double quotes to align with JS template and update example comment

cticenhour committed:
Restrict tag extension example to moose website build

cticenhour committed:
Make paths relative in tagging example page HTML

friedmud commented:
That can be easily changed.  Just a difference of `||` vs. `&&`

Is that a change you want _right now_?

cticenhour commented:
@GiudGiud I don't know if a tagview change is necessary for the scope of this PR. I could remove this example, and we can adjust behavior later in the VTB. I will defer to your opinion as reviewer, but wanted to share mine.

GiudGiud reviewed (approved):
looks good. Thanks for taking this over the finish line @cticenhour !

I ll add the tags to the VTB but we can't move forward on the vtb until the change of behavior is made @friedmud

cticenhour commented:
No problem - thanks for your review and comments!
